### ISSUE TITLE
Incorrect/inconsistent command line switch behaviour protocol selection s_client

### ISSUE THREAD

**Issue by seccubus** [openssl/openssl#1268]
What I had expected
====
* -ssl2 will include ssl2 as a potential protocol
* -no_ssl2 will exclude it as protocol
* similar behaviour for -ssl3 -tls1 -tls1_1 -tls1_2 and -no_ssl3 -no_tls1 -no_tls1_1 -no_tls1_2
* if a positive switch is used, all others are asumed nagative
* if just a negative switch is used all others are assumed to be positve

E.g.:
`openssl s_client -no_tl1 -no_tls1_1 -no_tls1_2 ` should allow me to connect to an ssl2 and sslv3 service.

What happens
===========

Turning off TLS switches off ssl2 too (and sslv3 a bit)
-------------------------------------
When used against `openssl s_server -ssl2 -key key.pem -cert cert.pem`
```
$ bin/openssl.Darwin.x86_64 s_client -connect localhost:4433 -no_tls1 -no_tls1_1 -no_tls1_2
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
CONNECTED(00000003)
140735180640336:error:140790E5:SSL routines:ssl23_write:ssl handshake failure:s23_lib.c:177:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 169 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1467202817
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
```

Against server: `$ bin/openssl.Darwin.x86_64 s_server -ssl3 -key key.pem -cert cert.pem`

Client side output:
```
$ bin/openssl.Darwin.x86_64 s_client -connect localhost:4433 -no_tls1 -no_tls1_1 -no_tls1_2
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
CONNECTED(00000003)
depth=0 C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
verify error:num=18:self signed certificate
verify return:1
depth=0 C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
verify return:1
140735180640336:error:060A209F:digital envelope routines:EVP_MD_size:message digest is null:evp_lib.c:312:
---
Certificate chain
 0 s:/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
   i:/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDtTCCAp2gAwIBAgIJALMOYpzO+8ZSMA0GCSqGSIb3DQEBBQUAMEUxCzAJBgNV
BAYTAkFVMRMwEQYDVQQIEwpTb21lLVN0YXRlMSEwHwYDVQQKExhJbnRlcm5ldCBX
aWRnaXRzIFB0eSBMdGQwHhcNMTYwNjI5MTIxNzM2WhcNMTYwNzA0MTIxNzM2WjBF
MQswCQYDVQQGEwJBVTETMBEGA1UECBMKU29tZS1TdGF0ZTEhMB8GA1UEChMYSW50
ZXJuZXQgV2lkZ2l0cyBQdHkgTHRkMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIB
CgKCAQEAvPYul2ItjCpUeXkAjJonAtLREBEZ7nWkmK8EZinIkqS15uYdZkXbNWqH
TS8OL9n6tr266iv6Aw5rIfhvGVnunY1OyKhj+7JjwTm4bGWfLLk9OnqxoXmisNle
bNOwwmI1nhEcQuyOS0DFuVUMYZN2Cq6hgFES8Sj7ABTcpA6FuCKA36JDfqEBqOXY
KURYNEL5T4j0qIUw7YuCplxZHK+6WZb68GZ1FPDtR3FWi49muw84SbOnY67G6URc
JyBbcAcG9TCE2H7juzAUBaBrH0g6w9dEgdZERwfAW3vaqUIj7elLZXjVKZHqQiCf
D2cDGqdDJBjEL04PHYpNGR0BqchkwwIDAQABo4GnMIGkMB0GA1UdDgQWBBS8DaKE
Erno12huALXeCyRzSrivZTB1BgNVHSMEbjBsgBS8DaKEErno12huALXeCyRzSriv
ZaFJpEcwRTELMAkGA1UEBhMCQVUxEzARBgNVBAgTClNvbWUtU3RhdGUxITAfBgNV
BAoTGEludGVybmV0IFdpZGdpdHMgUHR5IEx0ZIIJALMOYpzO+8ZSMAwGA1UdEwQF
MAMBAf8wDQYJKoZIhvcNAQEFBQADggEBAJKp7o4SQliLAJBIwIcbQ5NXbVMVOnZC
BoOZlFJ/lb6z63jZpjd+gmGVCHeaO6tYM1QpNXNf0Uo4SgsyHil2g/YmkczMU85e
kIl8x4GS+XiHhFuiNltKl5K+9Ky3OOmtI+PI1iBjUahbIMWGExNfeAhovsJfJp3+
fZKnPLIiwGXsLMqo7+OfnmhGubgWRDt3x5JKODNmLuvprF0XtkD9yRVPIxk1ckal
Jur4KNSixP+yC8FpLwVyDWBEOUxO489eDIBPAxQSDamQaTj0volQLqwuKqZfwYDZ
7wRkqZ5Z3yatWaw6W2szNndmjr600102SJ5f4FqfqtMpy15hlBArxH8=
-----END CERTIFICATE-----
subject=/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
issuer=/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
---
No client certificate CA names sent
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 1399 bytes and written 81 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-SHA
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-SHA
    Session-ID: F842058528D597846AABF167AA93463AC89A89572970AEF6394120747F7CCA51
    Session-ID-ctx:
    Master-Key: EF98D3A1796C3EA16AAD20EFDC4AE14E96488BF3DB2F4AF2A4AEBC4C232CC0AE699287FC65B2884C2EBFE4497437E027
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Compression: 1 (zlib compression)
    Start Time: 1467203545
    Timeout   : 300 (sec)
    Verify return code: 18 (self signed certificate)
---
$
```
Connect is actually closed.

server side output:
```
 bin/openssl.Darwin.x86_64 s_server -ssl3 -key key.pem -cert cert.pem
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
Using default temp DH parameters
ACCEPT
ERROR
shutting down SSL
CONNECTION CLOSED
ACCEPT
```

Only last tls protocol mentioned is used
------------------
Server: `$ bin/openssl.Darwin.x86_64 s_server -tls1 -key key.pem -cert cert.pem`

tls1_2 switch last
```
$ bin/openssl.Darwin.x86_64 s_client -connect localhost:4433 -no_ssl2 -no_ssl3 -tls1 -tls1_1 -tls1_2
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
CONNECTED(00000003)
140735180640336:error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number:s3_pkt.c:362:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 5 bytes and written 7 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1467203147
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
---
```

tls1.1 switch last
```
 bin/openssl.Darwin.x86_64 s_client -connect localhost:4433 -no_ssl2 -no_ssl3 -tls1 -tls1_2 -tls1_1
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
CONNECTED(00000003)
140735180640336:error:1408F10B:SSL routines:SSL3_GET_RECORD:wrong version number:s3_pkt.c:362:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 5 bytes and written 7 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.1
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1467203316
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
---
```

tls1.0 switch last:
```
$ bin/openssl.Darwin.x86_64 s_client -connect localhost:4433 -no_ssl2 -no_ssl3 -tls1_1 -tls1_2 -tls1
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
CONNECTED(00000003)
depth=0 C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
verify error:num=18:self signed certificate
verify return:1
depth=0 C = AU, ST = Some-State, O = Internet Widgits Pty Ltd
verify return:1
---
Certificate chain
 0 s:/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
   i:/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIDtTCCAp2gAwIBAgIJALMOYpzO+8ZSMA0GCSqGSIb3DQEBBQUAMEUxCzAJBgNV
BAYTAkFVMRMwEQYDVQQIEwpTb21lLVN0YXRlMSEwHwYDVQQKExhJbnRlcm5ldCBX
aWRnaXRzIFB0eSBMdGQwHhcNMTYwNjI5MTIxNzM2WhcNMTYwNzA0MTIxNzM2WjBF
MQswCQYDVQQGEwJBVTETMBEGA1UECBMKU29tZS1TdGF0ZTEhMB8GA1UEChMYSW50
ZXJuZXQgV2lkZ2l0cyBQdHkgTHRkMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIB
CgKCAQEAvPYul2ItjCpUeXkAjJonAtLREBEZ7nWkmK8EZinIkqS15uYdZkXbNWqH
TS8OL9n6tr266iv6Aw5rIfhvGVnunY1OyKhj+7JjwTm4bGWfLLk9OnqxoXmisNle
bNOwwmI1nhEcQuyOS0DFuVUMYZN2Cq6hgFES8Sj7ABTcpA6FuCKA36JDfqEBqOXY
KURYNEL5T4j0qIUw7YuCplxZHK+6WZb68GZ1FPDtR3FWi49muw84SbOnY67G6URc
JyBbcAcG9TCE2H7juzAUBaBrH0g6w9dEgdZERwfAW3vaqUIj7elLZXjVKZHqQiCf
D2cDGqdDJBjEL04PHYpNGR0BqchkwwIDAQABo4GnMIGkMB0GA1UdDgQWBBS8DaKE
Erno12huALXeCyRzSrivZTB1BgNVHSMEbjBsgBS8DaKEErno12huALXeCyRzSriv
ZaFJpEcwRTELMAkGA1UEBhMCQVUxEzARBgNVBAgTClNvbWUtU3RhdGUxITAfBgNV
BAoTGEludGVybmV0IFdpZGdpdHMgUHR5IEx0ZIIJALMOYpzO+8ZSMAwGA1UdEwQF
MAMBAf8wDQYJKoZIhvcNAQEFBQADggEBAJKp7o4SQliLAJBIwIcbQ5NXbVMVOnZC
BoOZlFJ/lb6z63jZpjd+gmGVCHeaO6tYM1QpNXNf0Uo4SgsyHil2g/YmkczMU85e
kIl8x4GS+XiHhFuiNltKl5K+9Ky3OOmtI+PI1iBjUahbIMWGExNfeAhovsJfJp3+
fZKnPLIiwGXsLMqo7+OfnmhGubgWRDt3x5JKODNmLuvprF0XtkD9yRVPIxk1ckal
Jur4KNSixP+yC8FpLwVyDWBEOUxO489eDIBPAxQSDamQaTj0volQLqwuKqZfwYDZ
7wRkqZ5Z3yatWaw6W2szNndmjr600102SJ5f4FqfqtMpy15hlBArxH8=
-----END CERTIFICATE-----
subject=/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
issuer=/C=AU/ST=Some-State/O=Internet Widgits Pty Ltd
---
No client certificate CA names sent
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 1618 bytes and written 354 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-SHA
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: zlib compression
Expansion: zlib compression
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1
    Cipher    : ECDHE-RSA-AES256-SHA
    Session-ID: 65E995BD2EC54E32698F0C2B933CB1F442D57E9E963724F32721FAD201620B54
    Session-ID-ctx:
    Master-Key: 5E62007B1E450998C1646B3284569DDC1D470642D4B0A519FFF68324DD2E07641A006BBCED9869681079892C26D0CFB0
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 7200 (seconds)
    TLS session ticket:
    0000 - db 57 71 12 d5 10 75 a2-34 73 b5 8b 29 34 03 68   .Wq...u.4s..)4.h
    0010 - 97 6e f7 ee 86 c8 27 8b-88 a0 a8 45 30 1c e8 42   .n....'....E0..B
    0020 - f2 94 d4 e1 9d b7 1b 37-ed 83 3e ec c5 ac 9b df   .......7..>.....
    0030 - 62 40 7a 87 41 5a 2c fd-36 cc 1e ce 59 29 c7 65   b@z.AZ,.6...Y).e
    0040 - e3 11 8f 26 68 98 07 06-8c bf 60 bf 08 cb a5 c7   ...&h.....`.....
    0050 - ba 8e c2 4c 6a d5 cf cf-11 51 6b 1e 9f 7c fe b4   ...Lj....Qk..|..
    0060 - 05 03 46 db fa 1b 69 18-9b 14 96 32 7b c0 f9 c7   ..F...i....2{...
    0070 - fb ca 3a 88 b9 08 4f ce-4b 7c 25 fa 9e 4c c9 16   ..:...O.K|%..L..
    0080 - 20 f1 54 6f 2a b6 1d 31-bd c1 f0 b9 02 7e d7 fd    .To*..1.....~..
    0090 - 6e 36 2e 9e 33 fc c0 74-f3 bf 0d 3f f6 38 d7 d1   n6..3..t...?.8..

    Compression: 1 (zlib compression)
    Start Time: 1467203417
    Timeout   : 7200 (sec)
    Verify return code: 18 (self signed certificate)
---
```

**Comment by mattcaswell** [comment]
What version of OpenSSL are you using? Your expectation of what the switches do is not quite right. In particular the "postive" switches *force* that particular version, i.e. "-tls1_2" forces TLSv1.2 (and only TLSv1.2). It doesn't make sense to list more than one, or to use one at the same time as any of the "no" switches (so we should probably prevent that from occurring). I'm guessing that's why its only the last one that actually has an effect.

**Comment by seccubus** [comment]
This is the version I am using:
```
$ bin/openssl.Darwin.x86_64
WARNING: can't open config file: /usr/local/ssl/openssl.cnf
OpenSSL> version
OpenSSL 1.0.2-chacha (1.0.2e-dev)
```

The 'positive' switches indeed all follow the same behaviour as in whichever one you use last is actually the one that is chosen. 
However this doesn't explain that sslv3 and sslv3 connections cannot be made if the switches `-no_tls1 -no_tls1_1 -no_tls1_2` are added.

I'm currently maintaining the client simulation code for testssl.sh and encountered this odd behaviour. (Managed to work around it now, btw)

**Comment by levitte** [comment]
Regarding the ssl3, I'm seeing that error line at the end of server cert verification, and am wondering what that is about.  I just tried, using `apps/server.pem`, and it worked like a charm.

**Comment by mattcaswell** [comment]
I am not surprised that you are unable to create SSLv2 connections - that is expected behaviour. SSLv2 is quite different to SSLv3+ - the ClientHello format is completely different. For a while clients used to send a "SSLv2 compatible ClientHello" which enabled you to negotiate SSLv2 or SSLv3+ - but imposed various restrictions (like an inability to negotiate extensions). That is no longer common practice and SSLv3/TLS style ClientHellos are sent by all modern clients (and this has been the default since OpenSSL 1.0.0) - and this means that you can no longer negotiate SSLv2. You can force an SSLv2 compat ClientHello by including SSLv2 ciphersuites in the "-cipher" flag to s_client. You will have to similarly enable SSLv2 ciphersuites in s_server too.

I'm not sure why you cannot negotiate SSLv3. This works for me:
```
$ ./openssl s_server -ssl3
```
and
```
$ ./openssl s_client -no_tls1 -no_tls1_1 -no_tls1_2
```
I note that you are using a non-standard OpenSSL - "1.0.2-chacha" is not an OpenSSL project produced version. Have you tried standard OpenSSL?

**Comment by mattcaswell** [comment]
The issue "Only last tls protocol mentioned is used" has been fixed in both master (1.1.0) and 1.0.2 branches. The SSLv2 issue is expected as described above, and I cannot reproduce the SSLv3 issue. Closing this unless we can get further information as to why this doesn't work for you.
